initramfs/init.sh

```diff
@@ -0,0 +1,202 @@
+#!/bin/busybox sh
+#
+# Copyright (C) 2018 Ultimaker B.V.
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
+#
+# SPDX-License-Identifier: AGPL-3.0+
+
+set -eu
+
+ROOT_MOUNT="/mnt/root"
+UPDATE_MOUNT="/mnt/update"
+UPDATE_SCRIPT="${UPDATE_MOUNT}/um_update.sh"
+UPDATE_DEVICES="/dev/sd[a-z][0-9] /dev/mmcblk[0-9]p[0-9]"
+BB_BIN="/bin/busybox"
+
+init="/sbin/init"
+ro=""
+root=""
+rootflags=""
+rootfstype="auto"
+
+shutdown()
+{
+	while [ 1 ]; do
+		poweroff
+		echo "Please remove power to complete shutdown."
+		sleep 10s
+	done
+}
+
+restart()
+{
+	echo "Rebooting in 5 seconds ..."
+	sleep 5s
+	reboot
+	modprobe sunxi_wdt || true
+	watchdog -T 1 -t 60 -F /dev/watchdog
+	shutdown
+}
+
+rescue_shell()
+{
+	set +eu
+	${BB_BIN} echo -en "\n" \
+		"##################################################\n" \
+		"#                   ▄▄▄      ▄▄▄                 #\n" \
+		"#                   ███████▄▄███▄                #\n" \
+		"#                 ▄███  ▄▄   ▄▄ ▐█               #\n" \
+		"#              ▄██  ██ ▀▀▀   ▀▀▀▐█▄              #\n" \
+		"#              ███▄▄██  ▄▄▄▄▄▄▄ ▐██              #\n" \
+		"#               ██████ ▀       ▀ ▐█              #\n" \
+		"#            ▄████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█▄           #\n" \
+		"#          ████▐█ ▐████████████████▌▐██▄         #\n" \
+		"#        ▐███▌▄▄█ ▐████████████████▌▐█▄█         #\n" \
+		"#         ███████ ███████▀▀ ▀██████▌▐██▌         #\n" \
+		"#         ███████ █████████ ██▐████▌▐██▌         #\n" \
+		"#        ▄███████ ██████▄██████████▌▐█▀█         #\n" \
+		"#        ████  ▐█ ▐████████████████▌▐█ █▌        #\n" \
+		"#        ▀███ ███   ▀▀▀▀            ▐██▀         #\n" \
+		"#           ▀▀████▄▄████████████████▀            #\n" \
+		"#             ▐████████▀▀████▀▀▌  █▌             #\n" \
+		"#            ▄██████  █  ▐███▄▄█▄▄██             #\n" \
+		"#            ████████▀▀▀█▀████▀    ██            #\n" \
+		"#            █████████    ▐███     ██            #\n" \
+		"#              ▀███▄█▄▄▄▄▄███▀▀▀▀▀▀▀             #\n" \
+		"#                                                #\n" \
+		"# Starting busybox rescue and recovery shell.    #\n" \
+		"# Tip: type help<enter> for available commands.  #\n" \
+		"#                                                #\n" \
+		"##################################################\n"
+	${BB_BIN} sh
+}
+
+critical_error()
+{
+	echo "A critical error has occurred, entering recovery mode."
+	rescue_shell
+	shutdown
+}
+
+boot_root()
+{
+	echo "Mounting ${root}."
+	mount -t ${rootfstype} -o exec,suid,dev,noatime,$rootflags,$ro "${root}" "${ROOT_MOUNT}"
+	kernel_umount
+	echo "Starting linux on ${root} of type ${rootfstype} with init=${init}."
+	exec switch_root /mnt/root "${init}"
```

Comment:
I see that in the calling code the "critical_error" is thrown when in case of an error the call to switch root returns with an error. But it would be more clear to handle the error more explicitly.

Reply:
I'm not sure if we can if-guard the exec switch root, as we are expecting to never return from it. If it returns, we get the critical error, but is it exec that fails? or is it switch root that fails. I think exec would return 0, as that did its work just fine, it was switch root that failed. Then again, it would seem logical that exec would just pass any return code along.
If that would be the case we can do:
'''sh
if exec switch_root /mnt/root "${init}"; then
    echo Failed to switch root
    exit 1
fi
'''

I'll try and update if this works.

Reply:
turns out, this doesn't work :)

exec will already switch to switch_root, so switch_root will be PID1 and the only active process. If it returns (which should not happen normally) it thus returns $ret to the kernel, which results in a kernel panic (attempting to kill(exit) init) which is not allowed.